opal/mca/installdirs/runtime/opal_installdirs_runtime.c

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2006-2007 Los Alamos National Security, LLC.  All rights
+ *                         reserved.
+ * Copyright (c) 2007      Cisco Systems, Inc.  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "opal_config.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "opal/constants.h"
+#include "opal/mca/installdirs/installdirs.h"
+#include "opal/runtime/opal.h"
+
+static int installdirs_runtime_open(void);
+
+opal_installdirs_base_component_t mca_installdirs_runtime_component = {
+    /* First, the mca_component_t struct containing meta information
+       about the component itself */
+    {OPAL_INSTALLDIRS_BASE_VERSION_2_0_0,
+
+     /* Component name and version */
+     "runtime", OPAL_MAJOR_VERSION, OPAL_MINOR_VERSION, OPAL_RELEASE_VERSION,
+
+     /* Component open and close functions */
+     installdirs_runtime_open, NULL},
+    {/* This component is checkpointable */
+     MCA_BASE_METADATA_PARAM_CHECKPOINT},
+
+    /* Next the opal_install_dirs_t install_dirs_data information */
+    {
+        NULL,
+    },
+};
+
+#include <dlfcn.h>
+#include "opal/util/basename.h"
+
+static int installdirs_runtime_open(void)
+{
+    Dl_info info;
+    void* opal_fct;
+
+    /* Casting from void* to fct pointer according to POSIX.1-2001 and POSIX.1-2008 */
+    *(void **)&opal_fct = dlsym(RTLD_DEFAULT, "opal_init_util");
+
+    if( 0 == dladdr(opal_fct, &info) ) {
+        /* Can't find the symbol */
+        return OPAL_ERROR;
+    }
+
+    char* dname = opal_dirname(info.dli_fname);
+    char* prefix = opal_dirname(dname);
+    free(dname);
+
+    mca_installdirs_runtime_component.install_dirs_data.prefix = prefix;
```

Comment:
Is there ordering / priority in the installdirs framework such that you can guarantee that this is the last component to be opened?  I.e., that writing this prefix will be the one that "wins"?

Reply:
This is not the last component, the last component is the original config. I put the priority of this to be inbetween the environment prefix and the static config data. However, both the env prefix and this module only fill the install prefix, everything else is statically set (as templates) by the config, which therefore must always be last.

There are still setups that don't make sense. As an example if the user manually set a prefix and that prefix is not the same as the loaded library. Is this bad or something the user purposely did (maybe because heterogeneous runs) ?

Reply:
Also, you might need to reset `exec_prefix`.  You'll need to see if the configured `prefix` and `exec_prefix` are the same, I think...?

Reply:
No need to change the `exec_prefix` as it it defined as `${prefix}` in the config component, and the templating logic will automatically set it.

Reply:
`exec_prefix` *defaults* to `${prefix}`, but it doesn't have to be.

```
./configure --prefix=/opt/openmpi --exec-prefix=/something/else ...
```

Reply:
I can't change the `exec_prefix` I have no idea where the binaries are executed if the user did what you suggested above. I feel like we're trying to solve all problems here, that was not the goal of this PR. The goal was to simply let people move their install directory and have things work properly.

Reply:
`libdir` defaults to `${exec_prefix}/lib`.  Hence, if `exec_prefix` is not the same as `prefix`, and `libdir` is set relative to `exec_prefix`, then setting `install_dirs_data.prefix` won't do what you expect.

We're not trying to solve everything here -- but solving this problem does have to be done correctly / handle the use cases that already exist.